libp2p/Host.cpp

```diff
@@ -367,22 +372,233 @@ void Host::runAcceptor()
 	}
 }
 
-void Host::doHandshake(bi::tcp::socket* _socket, NodeId _nodeId)
+void PeerHandshake::transition(boost::system::error_code _ech)
 {
-	try {
-		clog(NetConnect) << "Accepting connection for " << _socket->remote_endpoint();
-	} catch (...){}
-
-	shared_ptr<Peer> p;
-	if (_nodeId)
-		p = m_peers[_nodeId];
+	if (_ech)
+	{
+		boost::system::error_code ec;
+		socket->shutdown(boost::asio::ip::tcp::socket::shutdown_both, ec);
+		if (socket->is_open())
+			socket->close();
+		return;
+	}
 	
-	if (!p)
-		p.reset(new Peer());
-	p->endpoint.tcp.address(_socket->remote_endpoint().address());
+	auto self(shared_from_this());
+	if (nextState == New)
+	{
+		nextState = AckAuth;
+		
+		clog(NetConnect) << "Authenticating connection for " << socket->remote_endpoint();
+		
+		if (originated)
+		{
+			clog(NetConnect) << "devp2p.connect.egress sending auth";
+			// egress: tx auth
+			asserts((bool)remote);
+			auth.resize(Signature::size + h256::size + Public::size + h256::size + 1);
+			bytesRef sig(&auth[0], Signature::size);
+			bytesRef hepubk(&auth[Signature::size], h256::size);
+			bytesRef pubk(&auth[Signature::size + h256::size], Public::size);
+			bytesRef nonce(&auth[Signature::size + h256::size + Public::size], h256::size);
+			
+			// E(remote-pubk, S(ecdhe-random, ecdh-shared-secret^nonce) || H(ecdhe-random-pubk) || pubk || nonce || 0x0)
+			crypto::ecdh::agree(host->m_alias.sec(), remote, ss);
+			sign(ecdhe.seckey(), ss ^ this->nonce).ref().copyTo(sig);
+			sha3(ecdhe.pubkey().ref(), hepubk);
+			host->m_alias.pub().ref().copyTo(pubk);
+			this->nonce.ref().copyTo(nonce);
+			auth[auth.size() - 1] = 0x0;
+			encryptECIES(remote, &auth, authCipher);
+			
+			ba::async_write(*socket, ba::buffer(authCipher), [this, self](boost::system::error_code ec, std::size_t)
+			{
+				transition(ec);
+			});
+		}
+		else
+		{
+			clog(NetConnect) << "devp2p.connect.ingress recving auth";
+			// ingress: rx auth
+			authCipher.resize(321);
+			ba::async_read(*socket, ba::buffer(authCipher, 321), [this, self](boost::system::error_code ec, std::size_t)
```

Comment:
Do I understand it correctly that `safe` is a Münchhausen-variable, i.e. it keeps `this` alive as long as the async callback is present?